dali/kernels/signal/window/extract_windows_cpu_test.cc

```diff
@@ -0,0 +1,192 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include <tuple>
+#include <vector>
+#include <complex>
+#include <cmath>
+#include "dali/kernels/scratch.h"
+#include "dali/kernels/signal/signal_kernel_utils.h"
+#include "dali/kernels/signal/window/extract_windows_cpu.h"
+#include "dali/test/test_tensors.h"
+#include "dali/test/tensor_test_utils.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+namespace window {
+namespace test {
+
+class ExtractWindowsCpuTest : public::testing::TestWithParam<
+  std::tuple<std::array<int64_t, 2>, int64_t, int64_t, int64_t, bool, bool>> {
+ public:
+  ExtractWindowsCpuTest()
+    : data_shape_(std::get<0>(GetParam()))
+    , window_length_(std::get<1>(GetParam()))
+    , window_step_(std::get<2>(GetParam()))
+    , in_time_axis_(std::get<3>(GetParam()))
+    , center_windows_(std::get<4>(GetParam()))
+    , reflect_pad_(std::get<5>(GetParam()))
+    , data_(volume(data_shape_))
+    , in_view_(data_.data(), data_shape_) {}
+
+  ~ExtractWindowsCpuTest() override = default;
+
+ protected:
+  void SetUp() final {
+    SequentialFill(in_view_, 0);
+  }
+  TensorShape<2> data_shape_;
+  int64_t window_length_ = -1, window_step_ = -1, in_time_axis_ = -1;
+  bool center_windows_ = true, reflect_pad_ = false;
+  std::vector<float> data_;
+  OutTensorCPU<float, 2> in_view_;
+};
+
+template <typename T>
+void print_data(const OutTensorCPU<T, 2>& data_view) {
+  auto sh = data_view.shape;
+  for (int i0 = 0; i0 < sh[0]; i0++) {
+    for (int i1 = 0; i1 < sh[1]; i1++) {
+      int k = i0 * sh[1] + i1;
+      LOG_LINE << " " << data_view.data[k];
+    }
+    LOG_LINE << "\n";
+  }
+}
+
+template <typename T>
+void print_data(const OutTensorCPU<T, 3>& data_view) {
+  auto sh = data_view.shape;
+  for (int i0 = 0; i0 < sh[0]; i0++) {
+    for (int i1 = 0; i1 < sh[1]; i1++) {
+      for (int i2 = 0; i2 < sh[2]; i2++) {
+        int k = i0 * sh[1] * sh[2] + i1 * sh[2] + i2;
+        LOG_LINE << " " << data_view.data[k];
+      }
+      LOG_LINE << "\n";
+    }
+    LOG_LINE << "\n";
+  }
+  LOG_LINE << "\n";
+}
+
+TEST_P(ExtractWindowsCpuTest, ExtractWindowsTest) {
+  using OutputType = float;
+  using InputType = float;
+  constexpr int Dims = 2;
+  constexpr int InputDims = Dims;
+  constexpr int OutputDims = Dims + 1;
+
+  ExtractWindowsCpu<OutputType, InputType, Dims> kernel;
+  check_kernel<decltype(kernel)>();
+
+  KernelContext ctx;
+  ExtractWindowsArgs args;
+  args.window_length = window_length_;
+  args.window_step = window_step_;
+  args.in_time_axis = in_time_axis_;
+  args.center_windows = center_windows_;
+  args.reflect_pad = reflect_pad_;
+
+  // Hamming window
+  //  float a0 = 0.54;
+  //  float a1 = 1.0f - a0;
+  std::vector<float> window_fn_data(window_length_);
+  for (int t = 0; t < window_length_; t++) {
+    window_fn_data[t] = 1.0f;  // a0 - a1 * cos(2.0f * M_PI * t / window_length_);
+  }
```

Comment:
Remove the comments about Hamming or use it. BTW, the sample centering problem applies (t + 0.5, as we've done in spectrogram).

Reply:
done